### ISSUE TITLE
Type Hints and Annotations

### ISSUE THREAD

**Issue by marcharper** [Axelrod-Python/Axelrod#808]
Now that we're Python 3.5+ exclusive, it's time for [type hints](https://docs.python.org/3/library/typing.html)! 

Note to contributors: please run mypy on any newly annotated files. After installing with `pip install mypy`, please run:

`mypy --ignore-missing-imports --follow-imports skip axelrod/<file.py>`

for whichever files you annotate. You should receive no errors. We'll get this added to the automated checks soon. Thanks!

Finally, note that we're officially supporting Python 3.5 and 3.6, so please don't use any of the newer typing features in Python 3.6 yet.

**Comment by souravsingh** [comment]
I am interested in working on the issue.

**Comment by marcharper** [comment]
@souravsingh Great! It's a huge undertaking so feel free to tackle it in chunks, maybe file by file, and open multiple PRs if you'd like.

A few notes:
* Start with the easy cases, things that are clearly ints or floats, things that are strings and unlikely to be anything else (like player names)
* Go inside -> out, meaning start with the basic things like actions and players rather than the derived things like actual strategies, to keep from rewriting when you change your mind
* Consider using type variables or `NewType` for things like `axelrod.Actions.C` (rather than say `str`) and then import that type as needed. This is so if we decide to e.g. change the actions to 0 and 1 rather than `'C'` and `'D'` then we'll get the rest of the types for free. E.g. the history would be a list of whatever type the action is, and `List[T]` where `T` is the type variable for action is better than `List[str]` as the type for history.

We're in uncharted territory here so if you are not sure how to annotate something, either leave it alone or open an issue and we'll discuss it.

**Comment by marcharper** [comment]
@meatballs @drvinceknight Have you seen [PEP 526](https://www.python.org/dev/peps/pep-0526/)? It's Python 3.6 only and relevant to this issue. Unfortunately Python 3.7 won't be released for 18 months...

**Comment by drvinceknight** [comment]
Just read it :)

**Comment by marcharper** [comment]
@drvinceknight we should look into running a type checker (with travis?) once we have a some annotations in.

**Comment by drvinceknight** [comment]
Yeah: great idea! Do you know of a type checker for Python? A quick google has lead me to give up as the wifi is a bit slow in my hotel right now...

**Comment by souravsingh** [comment]
@drvinceknight @marcharper We can make use of [mypy](http://www.mypy-lang.org/) for type checks. Guido likes the tool himself.

**Comment by marcharper** [comment]
`mypy` is probably a good choice, assuming we can make it work with travis  -- it's been around for a while (and is the only type checker I've used before).

**Comment by souravsingh** [comment]
There is pytest plugin for mypy here- https://github.com/dbader/pytest-mypy
We can utilise it

**Comment by marcharper** [comment]
Here's a [patch](https://github.com/zulip/zulip/commit/c80f6993211011fc995270114f51d38a2f0489c0) that another project used to incorporate `mypy`.

**Comment by souravsingh** [comment]
There is another package for type checking called Traits- https://github.com/enthought/traits

**Comment by drvinceknight** [comment]
> Here's a patch that another project used to incorporate mypy.

Looks simple enough to me.

> There is another package for type checking called Traits- https://github.com/enthought/traits

@souravsingh do you think there's anything that Traits adds over mypy?

**Comment by souravsingh** [comment]
@drvinceknight All I can think of is it supports Type Validation, Notification to other programs if the type of a varibale is changed. I haven't really used traits a lot.

**Comment by marcharper** [comment]
Thanks @juliahou8 and @souravsingh for getting the ball rolling on this issue!

**Comment by janga1997** [comment]
@marcharper I would like to work on this issue.

**Comment by marcharper** [comment]
Whenever possible other than tests IMO.

**Comment by MariosZoulias** [comment]
Hey.

I have heard that this issue isvery easy to begin with and am really beginner so i would like to start with it.
Someone suggested me to take this issue and add type hints on these files : finite_state_machines.py, hmm.py, human.py 

**I will start from (really) Zero** as i dont even know what is PRs and Type Hints , so i hope to have finished all these by Friday ...

So may i take this issue for this files ???

Thank you very much
Marios Zoulias

**Comment by souravsingh** [comment]
@MariosZoulias I would suggest you to work on any one of the file for the PR.
For contributing see here- http://axelrod.readthedocs.io/en/latest/tutorials/contributing/strategy/instructions.html
For Type hinting, Read this- https://docs.python.org/3/library/typing.html

**Comment by MariosZoulias** [comment]
Ok then i would like to work on this issue and start with hmm.py 
I didnt answer fast because i was searching about what is type hinting and more about it...
Can you give me some examples of file "before and after" type hinting in order to have a better view???
Thank you 
Marios Zoulias

**Comment by janga1997** [comment]
@MariosZoulias you can check out any of the merged PRs referencing this issue a few comments above above, and click on `Files Changed`.

**Comment by MariosZoulias** [comment]
Thank you very much for your fast aswer.
So anyway i have the issue about the hmm.py file ??? (am just beginner and i dont really know the procedure) (of course i read the contributing instructions)

**Comment by janga1997** [comment]
@MarioZoulias Could we take this up in the gitter room?
Also, i didn't quite understand your problem.

**Comment by MariosZoulias** [comment]
Yes of course am already there ;)

**Comment by souravsingh** [comment]
Here is the list of files in main module done so far-
- [ ] _strategy_utils.py
- [ ] actions.py
- [ ] deterministic_cache.py
- [ ] ecosystem.py
- [ ] eigen.py
- [ ] fingerprint.py  
- [ ] game.py
- [ ] graph.py
- [ ] interaction_utils.py
- [ ] load_data.py
- [ ] match.py
- [ ] match_generator.py
- [ ] mock_player.py
- [ ] moran.py
- [ ] player.py
- [ ] plot.py
- [ ]  random_.py
- [ ] result_set.py
- [ ] strategy_transformers.py
- [ ] tournament.py

**Comment by drvinceknight** [comment]
That's awesome :+1: @souravsingh

**Comment by souravsingh** [comment]
Let me know if I have missed something in the main module, I just skimmed through the code.

I will also be making a second list for the type hint progress in strategies module.

**Comment by janga1997** [comment]
I would like to take up `match.py` and `match_generator.py`, if that's fine.

**Comment by drvinceknight** [comment]
That's great, thanks @janga1997!

**Comment by souravsingh** [comment]
For strategies we have this list-

- [ ] _filters.py
- [ ] adaptive.py
- [ ] alternator.py
- [ ] ann.py
- [ ] apavlov.py
- [ ] alternator.py
- [ ] appeaser.py
- [ ] averagecopier.py
- [ ] axelrod_first.py
- [ ] axelrod_second.py 
- [ ] backstabber.py
- [ ] better_and_better.py
- [ ] calculator.py
- [ ] cooperator.py
- [ ] cycler.py
- [ ] darwin.py
- [ ] dbs.py
- [ ] doubler.py
- [ ] defector.py
- [ ] finite_state_machines.py
- [ ] forgiver.py
- [ ] gambler.py
- [ ] geller.py
- [ ] gobymajority.py
- [ ] gradualkiller.py
- [ ] grudger.py
- [ ] grumpy.py

- [ ] handshake.py
- [ ] hmm.py
- [ ] human.py
- [ ] hunter.py
- [ ] inverse.py
- [ ] lookerup.py
- [ ] mathematicalconstants.py
- [ ] memoryone.py
- [ ] memorytwo.py
- [ ] meta.py
- [ ] mindcontrol.py
- [ ] mindreader.py
- [ ] mutual.py
- [ ] negation.py
 - [ ] oncebitten.py
- [ ] prober.py
- [ ] punisher.py
- [ ] qlearner.py
- [ ] rand.py
- [ ] resurrection.py 
- [ ] retaliate.py
- [ ] selfesteem.py
- [ ] sequence_player.py
- [ ] stalker.py
- [ ] titfortat.py
- [ ] very_bad.py
- [ ] worse_and_worse.py

**Comment by eshansingh** [comment]
@souravsingh I'm looking to complete the whole thing. I'm taking up `graph.py` for now.

**Comment by souravsingh** [comment]
@eshansingh Awesome. Go for it

**Comment by pujaniceman** [comment]
are all the files completed here?

**Comment by drvinceknight** [comment]
They pretty much are, if there's a file that is said to be complete here but it doesn't look like it is let us know :)
